example/tests/test_serializers.py

```diff
@@ -193,3 +199,48 @@ def test_model_serializer_with_implicit_fields(self, comment, client):
 
         assert response.status_code == 200
         assert expected == response.json()
+
+
+class TestPolymorphicModelSerializer(TestCase):
+    def setUp(self):
+        self.project = ArtProjectFactory.create()
+
+    def test_polymorphic_model_serializer_passes_instance_to_child(self):
+        """
+        Ensure that `PolymorphicModelSerializer` is passing the instance to the
+        child serializer when initializing them
+        """
+        # Initialize a serializer that would partially update a model instance
+        initial_data = {"artist": "Mark Bishop", "type": "artProjects"}
+        parent_serializer = ProjectSerializer(
+            instance=self.project, data=initial_data, partial=True
+        )
+
+        # Override `__init__` with our own method
+        def init_with_asserts(child_self, instance=None, data=empty, **kwargs):
+            """
+            Override `ArtProjectSerializer.__init__` with the same signature that
+            `BaseSerializer.__init__` has to assert that it receives the parameters
+            that `BaseSerializer` expects
+            """
+            assert instance == self.project
+            assert data == initial_data
+            assert kwargs.get("partial", False) == parent_serializer.partial
+            assert kwargs.get("context", {}) == parent_serializer.context
+
+            return super(ArtProjectSerializer, child_self).__init__(
+                instance, data, **kwargs
+            )
+
+        original_init = ArtProjectSerializer.__init__
+        ArtProjectSerializer.__init__ = init_with_asserts
+
+        parent_serializer.is_valid(raise_exception=True)
+
+        # Run save to force `ProjectSerializer` to init `ArtProjectSerializer`
+        parent_serializer.save()
+
+        # Asserts in the overridden `__init__` method declared above
+
+        # Restore original init to avoid affecting other tests
+        ArtProjectSerializer.__init__ = original_init
```

Comment:
what when an assert fails? This will never be called and following test might therefore also fail.

Unfortunatly currently we have a mix between unit test and pytest styles in DJA. You could split it up that overwriting of init happens in setup and teardown but not sure this makes it more readable.

In a a case like this one using the newer pytest style would be much easier where you could use [monkeypatch](https://docs.pytest.org/en/latest/monkeypatch.html).

Reply:
Yes you are right, I did't think of that. Yes, pytest would be ideal and make a fixture or use monkeypatch, but I think we should keep it consistent with the rest of the file, so I'd go for using the unittest `tearDown`.

If I do the override of the init in the `setUp`, then I think I should move the asserts out of the overridden init and back into the test, using the instance of the test class to pass the args and kwargs from the init to the test.

The other option would be to keep the override in the test, but in the setup getting the reference to the original `ArtProjectSerializer.__init__` and in the tearDown restore it. So it would be something like:

```python
class TestPolymorphicModelSerializer(TestCase):
    def setUp(self):
        self.project = ArtProjectFactory.create()
        self.art_project_serializer_init = ArtProjectSerializer.__init__

    def tearDown(self):
        ArtProjectSerializer.__init__ = self.art_project_serializer_init

    def test_polymorphic_model_serializer_passes_instance_to_child(self):
        pass  # The rest of the stuff that's already there
```

What do you think?

Reply:
It is fine with me if you use unittest and move the asserts into the test method again.